File: Data Structures & Algorithms/climbing-stairs/submission-2.java
class Solution {
    public int climbStairs(int n) {
        
       int [] memo = new int[n + 1];
       return dp(n,memo);
    }

    public int dp(int n, int [] memo){
         //base case.
        if(n == 1) return 1;
        if(n == 2) return 2;

        if(memo[n] != 0){
            return memo[n];
        }
        //average case
        memo[n] = dp(n - 1, memo) + dp(n - 2, memo);
        return memo[n];
    }
}

File: Data Structures & Algorithms/duplicate-integer/submission-2.py
class Solution:
    def hasDuplicate(self, nums: List[int]) -> bool:
     #Approach 1. Brute force search.
        #base case if the list is empty.
        #if not nums:
          #  return False
        #base case if the list is just 1 element long.
        #if len(nums) == 1:
          #  return False
        
        #Average case 
        #for i in range(len(nums) - 1):
        # for j in range(i + 1, len(nums)):
         #  if nums[i]==nums[j]:
          #  return True
        #return False
    
     #Approach 2. Hashset
      hashSet = set()

      for n in nums:
       if n in hashSet:
        return True
       hashSet.add(n)
      return False

    
    